airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py

```diff
@@ -387,37 +276,59 @@ def next_page_token(self, response: requests.Response) -> Optional[Mapping[str,
+            yield from records_generator_fn(response, stream_state, stream_slice)
+
+            if not response:
+                pagination_complete = True
+            else:
+                next_page_token = self._next_page_token(response)
+                if not next_page_token:
+                    pagination_complete = True
+
+        # Always return an empty generator just in case no records were ever yielded
+        yield from []
+
     def read_records(
         self,
-        sync_mode: SyncMode,
-        cursor_field: Optional[List[str]] = None,
         stream_slice: Optional[StreamSlice] = None,
-        stream_state: Optional[StreamState] = None,
     ) -> Iterable[StreamData]:
-        # Warning: use self.state instead of the stream_state passed as argument!
         stream_slice = stream_slice or {}  # None-check
         # Fixing paginator types has a long tail of dependencies
-        self._paginator.reset()  # type: ignore
-        # Note: Adding the state per partition led to a difficult situation where the state for a partition is not the same as the
-        # stream_state. This means that if any class downstream wants to access the state, it would need to perform some kind of selection
-        # based on the partition. To short circuit this, we do the selection here which avoid downstream classes to know about it the
-        # partition. We have generified the problem to the stream slice instead of the partition because it is the level of abstraction
-        # streams know (they don't know about partitions). However, we're still unsure as how it will evolve since we can't see any other
-        # cursor doing selection per slice. We don't want to pollute the interface. Therefore, we will keep the `hasattr` hack for now.
-        # * What is the information we need to clean the hasattr? Once we will have another case where we need to select a state, we will
-        #    know if the abstraction using `stream_slice` so select to state is the right one and validate if the interface makes sense.
-        # * Why is this abstraction not on the DeclarativeStream level? DeclarativeStream does not have a notion of stream slicers and we
-        #    would like to avoid exposing the stream state outside of the cursor. This case is needed as of 2023-06-14 because of
-        #    interpolation.
-        if self.cursor and hasattr(self.cursor, "select_state"):  # type: ignore
-            slice_state = self.cursor.select_state(stream_slice)  # type: ignore
-        elif self.cursor:
-            slice_state = self.cursor.get_stream_state()
-        else:
-            slice_state = {}
+        self._paginator.reset()
 
         most_recent_record_from_slice = None
-        for stream_data in self._read_pages(self.parse_records, stream_slice, slice_state):
+        for stream_data in self._read_pages(self._parse_records, self.state, stream_slice):
```

Comment:
Without the hack, if a custom cursor is used with a per partition stream state and it relies on the parameter instead of their internal state, this will not work anymore. I know that we said they should use their internal state but that's not what GL used to do so I'm a bit wary of this

Reply:
Can you explain how a custom cursor could break here? How would it be called differently due to this change? To me it looks like no custom component will be called with a different object (the cursor never gets the slice passed in and as explained here: https://github.com/airbytehq/airbyte/pull/28657#discussion_r1274908937 no other component sees it either)